sql/src/main/java/io/crate/analyze/relations/RelationAnalysisContext.java

```diff
@@ -92,7 +91,9 @@ void addJoinType(JoinType joinType, @Nullable Symbol joinCondition) {
 
     List<JoinPair> joinPairs() {
         if (joinPairs == null) {
-            return ImmutableList.of();
+            // Must be mutable as new join pairs extracted
+            // from implicit join conditions might be added.
+            return new ArrayList<>();
```

Comment:
So far we've been working towards more immutability. I don't like that we're now going back the other way. Especially if it's not encapsulated.

List properties that are intentionally mutated are especially hard to reason about. Same with the `addToJoinCondition` change. I think we should try to re-work this in a way that avoids mutations across class boundaries.

Reply:
So, you think it's preferable to create a new JoinPair with the new condition instead?